libs/auth/src/certs.test.ts

```diff
@@ -0,0 +1,237 @@
+      electionHash: mockElectionHash,
+    },
+  },
+  {
+    subject:
+      'subject=C = US, ST = CA, O = VotingWorks, ' +
+      '1.3.6.1.4.1.59817.1 = card, ' +
+      `1.3.6.1.4.1.59817.2 = ${mockJurisdiction}, ` +
+      '1.3.6.1.4.1.59817.3 = pw, ' +
+      `1.3.6.1.4.1.59817.4 = ${mockElectionHash}`,
+    expectedUserData: {
+      role: 'poll_worker',
+      electionHash: mockElectionHash,
+    },
+  },
+])('parseUserDataFromCert', async ({ subject, expectedUserData }) => {
+  mockOf(openssl).mockImplementationOnce(() =>
+    Promise.resolve(Buffer.from(subject, 'utf-8'))
+  );
+  expect(await parseUserDataFromCert(mockCert, mockJurisdiction)).toEqual(
+    expectedUserData
+  );
+});
+
+test.each<{ description: string; subject: string }>([
+  {
+    description: 'machine cert instead of card cert',
+    subject:
+      'subject=C = US, ST = CA, O = VotingWorks, ' +
+      '1.3.6.1.4.1.59817.1 = admin, ' +
+      `1.3.6.1.4.1.59817.2 = ${mockJurisdiction}`,
+  },
+  {
+    description: 'wrong jurisdiction',
+    subject:
+      'subject=C = US, ST = CA, O = VotingWorks, ' +
+      '1.3.6.1.4.1.59817.1 = card, ' +
+      `1.3.6.1.4.1.59817.2 = ST.WrongJurisdiction, ` +
+      '1.3.6.1.4.1.59817.3 = sa',
+  },
+  {
+    description: 'missing card type',
+    subject:
+      'subject=C = US, ST = CA, O = VotingWorks, ' +
+      '1.3.6.1.4.1.59817.1 = card, ' +
+      `1.3.6.1.4.1.59817.2 = ${mockJurisdiction}`,
+  },
+  {
+    description: 'missing election hash for election card',
+    subject:
+      'subject=C = US, ST = CA, O = VotingWorks, ' +
+      '1.3.6.1.4.1.59817.1 = card, ' +
+      `1.3.6.1.4.1.59817.2 = ${mockJurisdiction}, ` +
+      '1.3.6.1.4.1.59817.3 = em',
+  },
+])('parseUserDataFromCert validation, $description', async ({ subject }) => {
+  mockOf(openssl).mockImplementationOnce(() =>
+    Promise.resolve(Buffer.from(subject, 'utf-8'))
+  );
+  await expect(
+    parseUserDataFromCert(mockCert, mockJurisdiction)
+  ).rejects.toThrow();
+});
+
+test.each<{
+  user: SystemAdministratorUser | ElectionManagerUser | PollWorkerUser;
+  expectedSubject: string;
+}>([
+  {
+    user: {
+      role: 'system_administrator',
+    },
+    expectedSubject:
+      '/C=US/ST=CA/O=VotingWorks' +
+      '/1.3.6.1.4.1.59817.1=card' +
+      `/1.3.6.1.4.1.59817.2=${mockJurisdiction}` +
+      '/1.3.6.1.4.1.59817.3=sa/',
+  },
+  {
+    user: {
+      role: 'election_manager',
+      electionHash: mockElectionHash,
+    },
+    expectedSubject:
+      '/C=US/ST=CA/O=VotingWorks' +
+      '/1.3.6.1.4.1.59817.1=card' +
+      `/1.3.6.1.4.1.59817.2=${mockJurisdiction}` +
+      '/1.3.6.1.4.1.59817.3=em' +
+      `/1.3.6.1.4.1.59817.4=${mockElectionHash}/`,
+  },
+  {
+    user: {
+      role: 'poll_worker',
+      electionHash: mockElectionHash,
+    },
+    expectedSubject:
+      '/C=US/ST=CA/O=VotingWorks' +
+      '/1.3.6.1.4.1.59817.1=card' +
+      `/1.3.6.1.4.1.59817.2=${mockJurisdiction}` +
+      '/1.3.6.1.4.1.59817.3=pw' +
```

Comment:
Is it useful to have the card type be a two-character abbreviation? Seems harder to read, so I was wondering if there's an intentional tradeoff

Reply:
Good q! I'd been following the plan laid out by Ben in [the certificates section of this doc](https://docs.google.com/document/d/1-6K-ayWUAIQAO3n50PZAAODVaXsBUkZhGoXcFNja6aI/edit#heading=h.j01awri45014) and hadn't questioned this. My initial read was that the two-character abbreviation was a space-saving measure, but that's truly inconsequential in the context of Java Cards, which give us so much more space than our old memory cards.

Ultimately, the decision doesn't matter too much since the we're not going to be working with the underlying card certificates other than in this file, but I'm very open to removing the abbreviations in favor of complete strings.

cc @benadida in case there are any other reasons for the user role abbreviations

Reply:
Going to merge this PR but will revisit this comment in a follow-up PR if need be!

Reply:
My goal was space-saving in cert, which makes a bigger difference than it might seem because cert format is highly compressed. That said, it's still not a huge difference, as @arsalansufi points out, so I'm game to change these.